### PR TITLE
Refactor get in source handler

### PR DESCRIPTION
**Pull request by mcoughlin**
This PR refactors the "get" in the source handler in preparation for getting sources from other handlers.

### CODE REVIEW

File: skyportal/handlers/api/source.py [hunk 1]
```diff
@@ -77,1919 +77,2067 @@
...
+    schema = Obj.__schema__()
+
+    ra = data.get('ra', None)
+    dec = data.get('dec', None)
+
+    if ra is None and not obj_already_exists:
+        raise AttributeError("RA must not be null for a new Obj")
+
+    if dec is None and not obj_already_exists:
+        raise AttributeError("Dec must not be null for a new Obj")
```
Comment (guynir42) [inline skyportal/handlers/api/source.py:1240]: Lets combine these into one conditional and one error message...

File: skyportal/handlers/api/source.py [hunk 2]
```diff
@@ -77,1919 +77,2067 @@
...
+
+    user_group_ids = [g.id for g in user.groups]
+    user_accessible_group_ids = [g.id for g in user.accessible_groups]
+    if not user_group_ids:
+        raise AttributeError(
+            "You must belong to one or more groups before " "you can add sources."
```
Comment (guynir42) [inline skyportal/handlers/api/source.py:1246]:
```suggestion
            "You must belong to one or more groups before you can add sources."
```

File: skyportal/handlers/api/source.py [hunk 3]
```diff
@@ -77,1919 +77,2067 @@
...
             nullable: true
             schema:
-              type: string
-            description: Comma separated string of origins. Only annotations from these origins are used when filtering with the annotationsFilter.
+              type: boolean
+            description: |
+              Boolean indicating whether to include the color-magnitude data from Gaia.
+              This will only include data for objects that have an annotation
+              with the appropriate format: a key named Gaia that contains a dictionary
+              with keys named Mag_G, Mag_Bp, Mag_Rp, and Plx
```
Comment (guynir42) [inline skyportal/handlers/api/source.py:1588]: I think this format has changed a bit when we added new catalogs.

**Review by guynir42 — APPROVED** [review]
I have some small comments. 
Thanks a lot for doing this, it was long overdue. 
I should only mention that we have a basic problem of filtering on e.g., comments after pagination/sorting is done, which is not trivial to solve at this point, and we should think about how we will deal with that.

File: skyportal/handlers/api/source.py [hunk 4]
```diff
@@ -77,1919 +77,2067 @@
...
-    else:
-        flow = Flow()
-        flow.push(
-            '*', "skyportal/REFRESH_SOURCE", payload={"obj_key": obj.internal_key}
-        )
-        flow.push('*', "skyportal/REFRESH_CANDIDATE", payload={"id": obj.internal_key})
+    user = session.query(User).get(user_id)
 
-    return obj.id
+    # Fetch multiple sources
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/source.py:391]: Added.

File: skyportal/handlers/api/source.py [hunk 5]
```diff
@@ -77,1919 +77,2067 @@
...
+        photstat_subquery = (
+            PhotStat.query_records_accessible_by(user)
+            .filter(PhotStat.first_detected_mjd >= Time(start_date).mjd)
+            .subquery()
         )
-    return query
+        obj_query = obj_query.join(
+            photstat_subquery, Obj.id == photstat_subquery.c.obj_id
+        )
+    if end_date:
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/source.py]: added.

[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/source.py]: added.

File: skyportal/handlers/api/source.py [hunk 6]
```diff
@@ -77,1919 +77,2067 @@
...
+            user
+        ).subquery()
+
+        nonclassification_query = (
             DBSession()
-            .query(Source)
-            .filter(Source.obj_id == obj_id)
-            .filter(Source.group_id.in_(user_group_ids))
-            .count()
+            .query(
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/source.py]: yes.

File: skyportal/handlers/api/source.py [hunk 7]
```diff
@@ -77,1919 +77,2067 @@
...
+        if isinstance(comments_filter_author, str) and "," in comments_filter_author:
+            comments_filter_author = [
+                c.strip() for c in comments_filter_author.split(",")
+            ]
+        elif isinstance(comments_filter_author, str):
+            comments_filter_author = [comments_filter_author]
+        else:
+            raise ValueError(
+                "Invalid commentsFilterAuthor value -- must provide at least one string value"
+            )
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/source.py]: This keeps it out of the loop. As you said, we have to think more about how to do this in the DB.

File: skyportal/handlers/api/source.py (hunk 1 above)
Comment (mcoughlin) [reply skyportal/handlers/api/source.py]: yes.

File: skyportal/handlers/api/source.py (hunk 3 above)
Comment (mcoughlin) [reply skyportal/handlers/api/source.py]: yup!

**Pull request merged**